File: algorithms/blind75/1/valid_parantheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        closing = {k: v for v, k in zip("{[(", "}])")}
        stack = []
        for c in s:
            if c in closing:
                if len(stack) == 0:
                    return False
                c = closing[c]
                if c != stack.pop():
                    return False
            else:
                stack.append(c)
        
        return len(stack) == 0
        

File: algorithms/blind75/1/valid_parantheses.cpp
class Solution {
public:
    bool isValid(string s) {
        // map<char, int> ctr     = {{'{', 0}, {'(', 0}, {'[', 0}};
        map<char, int> closing = {{'}', '{'}, {']', '['}, {')', '('}};
        char last_opened;
        vector<char> stack;
        for (char c : s) {
            if (closing.count(c)) {
                c = closing[c];
                if (!size(stack)) return false;
                last_opened = stack.back();
                stack.pop_back();
                if (c != last_opened)
                    return false;
            } else {
                stack.push_back(c);
            }
        }
        return size(stack) == 0;
    }
};
